core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -71,49 +71,46 @@ object FetchSession {
   * Note that fetcherLogStartOffset is the LSO of the follower performing the fetch, whereas
   * localLogStartOffset is the log start offset of the partition on this broker.
   */
-class CachedPartition(val topic: String,
-                      val partition: Int,
+class CachedPartition(val topicPartition: TopicPartition,
                       var maxBytes: Int,
                       var fetchOffset: Long,
                       var highWatermark: Long,
+                      var leaderEpoch: Optional[Integer],
                       var fetcherLogStartOffset: Long,
                       var localLogStartOffset: Long)
     extends ImplicitLinkedHashCollection.Element {
 
   var cachedNext: Int = ImplicitLinkedHashCollection.INVALID_INDEX
   var cachedPrev: Int = ImplicitLinkedHashCollection.INVALID_INDEX
 
-  override def next = cachedNext
-  override def setNext(next: Int) = this.cachedNext = next
-  override def prev = cachedPrev
-  override def setPrev(prev: Int) = this.cachedPrev = prev
+  override def next: Int = cachedNext
+  override def setNext(next: Int): Unit = this.cachedNext = next
+  override def prev: Int = cachedPrev
+  override def setPrev(prev: Int): Unit = this.cachedPrev = prev
 
   def this(topic: String, partition: Int) =
-    this(topic, partition, -1, -1, -1, -1, -1)
+    this(new TopicPartition(topic, partition), -1, -1, -1, Optional.empty(), -1, -1)
```

Comment:
The decision not to use `TopicPartition` was actually deliberate here.  TP has an extra "hash" field which we don't need, which would add 4 bytes of overhead per partition.  It also adds the overhead of a Java object, which is probably at least 8 bytes per partition.

Reply:
Ok, I'll revert this change and add a comment.

Reply:
There already was a comment (which I had ignored 😝 ).